### PR TITLE
[SkyServe] `sky serve` CLI prototype

### PR DESCRIPTION
**Pull request by cblmemo**
This PR introduces `sky serve` CLI for starting a service.

Currently supported CLI:
* `sky serve up <task_yaml>`
* `sky serve down <service_name>`
* `sky serve status (<service_name>)`

To see the log of control-plane, use `sky logs <controller_name> 1`;
To see the log of redirector, use `sky logs <controller_name> 2`.
You can find the name of controller by `sky serve status -a`.

TODO: Maybe have a `sky serve logs` CLI.

serve-cli-example

serve-result
(The IP shown in the browser is the actual replica's IP)

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Review by infwinston — COMMENTED** [review]
Thanks @cblmemo ! Just left some initial comments for first round review. also, just to confirm we will leave user app error handling in next PR right?

File: sky/setup_files/setup.py [hunk 1]
```diff
@@ -112,7 +112,10 @@ def parse_readme(readme: str) -> str:
     'pulp',
     # Ray job has an issue with pydantic>2.0.0, due to API changes of pydantic. See
     # https://github.com/ray-project/ray/issues/36990
-    'pydantic<2.0'
+    'pydantic<2.0',
+    # Required by the SkyServe library
+    'uvicorn',
+    'fastapi'
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/setup_files/setup.py:118]: One problem is we want to use this dependency and we only install cloud corresponding dependencies, for example, https://github.com/skypilot-org/skypilot/blob/c35a1cb901f8099ad1121f520cf74ba90ceae883/sky/templates/gcp-ray.yml.j2#L196. Shall we change the yml template instead?

File: sky/backends/cloud_vm_ray_backend.py [hunk 2]
```diff
@@ -2924,7 +2924,9 @@ def _exec_code_on_head(
                     '\nTo view the spot job dashboard:\t'
                     f'{backend_utils.BOLD}sky spot dashboard'
                     f'{backend_utils.RESET_BOLD}')
-            else:
+            elif not name.startswith('middleware-'):
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/backends/cloud_vm_ray_backend.py:2927]: This naming still has the problem that package name conflicts inside `serve` library.. If using this name, there will be two `autoscaler` packages inside the `serve` dir

**Comment by cblmemo** [comment]
Move one discussion in #2267 here: For replica_policy's upper and lower threshold, I'll just check if it is None for now, then if the user only specifies upper_threshold, that means the user only wants to autoscale up and #replica will only increase until reach max_replica; vise versa for only specify lower_threshold case. Only if the user specified both upper and lower thresholds will enable autoscale up and down. We might need a discussion on here.

**Comment by cblmemo** [comment]
TOOD: Add a section in `sky status` and disable middleware in sky status clusters

**Comment by cblmemo** [comment]
Merged and delay code review when merging to master branch

**Pull request merged**